### DISCUSSION TITLE
time complexity of TSNode::* APIs

### DISCUSSION THREAD

**Discussion by champignoom** [tree-sitter/tree-sitter#2250] [Q&A] [answered]
By reading the implementation, I found just recently that the APIs might be more expensive than the names suggest. For example, `TSNode::parent()` does not just return a cached pointer in O(1) but rather goes from the root all the way down to itself and checks parenthoods on the fly, which costs O(n) in the worst case if I'm not mistaken.
https://github.com/tree-sitter/tree-sitter/blob/a62bac5370dc5c76c75935834ef083457a6dd0e1/lib/src/node.c#L461-L490

see also the implementation of [`TSNode::next_sibling`](https://github.com/tree-sitter/tree-sitter/blob/a62bac5370dc5c76c75935834ef083457a6dd0e1/lib/src/node.c#L246C22-L295)

I'm wondering if there's any possible/planned optimization for this, and if not, will this be documented?

**Comment by ahlinc** [comment, marked as answer]
It's a known _feature_ and it's a cost of immutable trees with a cheap cloning operation what allows to work with the same trees in parallel from multiple threads. _It will be documented._

**Reply by ahlinc** [reply]
> which costs O(n) in the worst case if I'm not mistaken

You're right that it's not `O(1)` but it does pretty effective stepping down over a tree to a node by using checks of predecessor ranges.
